pkg/daemon/certificate_writer.go

```diff
@@ -101,92 +117,230 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
-				stillExists := false
-				for _, CA := range mergedData {
-					// if one of our spec CAs matches the existing file, we are good.
-					if CA.File == entry.Name() {
-						stillExists = true
+			for _, entry := range entries {
+				if entry.IsDir() {
+					stillExists := false
+					for _, CA := range mergedData {
+						// if one of our spec CAs matches the existing file, we are good.
+						if CA.File == entry.Name() {
+							stillExists = true
+						}
 					}
-				}
-				if !stillExists {
-					if err := os.RemoveAll(filepath.Join("/etc/docker/certs.d", entry.Name())); err != nil {
-						klog.Warningf("Could not remove old certificate: %s", filepath.Join("/etc/docker/certs.d", entry.Name()))
+					if !stillExists {
+						if err := os.RemoveAll(filepath.Join("/etc/docker/certs.d", entry.Name())); err != nil {
+							klog.Warningf("Could not remove old certificate: %s", filepath.Join("/etc/docker/certs.d", entry.Name()))
+						}
 					}
 				}
 			}
-		}
 
-		for _, CA := range controllerConfig.Spec.ImageRegistryBundleData {
-			caFile := strings.ReplaceAll(CA.File, "..", ":")
-			if err := os.MkdirAll(filepath.Join(imageCAFilePath, caFile), defaultDirectoryPermissions); err != nil {
-				return err
+			for _, CA := range controllerConfig.Spec.ImageRegistryBundleData {
+				caFile := strings.ReplaceAll(CA.File, "..", ":")
+				if err := os.MkdirAll(filepath.Join(imageCAFilePath, caFile), defaultDirectoryPermissions); err != nil {
+					return err
+				}
+				if err := writeFileAtomicallyWithDefaults(filepath.Join(imageCAFilePath, caFile, "ca.crt"), CA.Data); err != nil {
+					return err
+				}
 			}
-			if err := writeFileAtomicallyWithDefaults(filepath.Join(imageCAFilePath, caFile, "ca.crt"), CA.Data); err != nil {
-				return err
+
+			for _, CA := range controllerConfig.Spec.ImageRegistryBundleUserData {
+				caFile := strings.ReplaceAll(CA.File, "..", ":")
+				if err := os.MkdirAll(filepath.Join(imageCAFilePath, caFile), defaultDirectoryPermissions); err != nil {
+					return err
+				}
+				if err := writeFileAtomicallyWithDefaults(filepath.Join(imageCAFilePath, caFile, "ca.crt"), CA.Data); err != nil {
+					return err
+				}
 			}
+
 		}
+	}
 
-		for _, CA := range controllerConfig.Spec.ImageRegistryBundleUserData {
-			caFile := strings.ReplaceAll(CA.File, "..", ":")
-			if err := os.MkdirAll(filepath.Join(imageCAFilePath, caFile), defaultDirectoryPermissions); err != nil {
-				return err
-			}
-			if err := writeFileAtomicallyWithDefaults(filepath.Join(imageCAFilePath, caFile, "ca.crt"), CA.Data); err != nil {
-				return err
-			}
+	oldAnno := dn.node.Annotations[constants.ControllerConfigSyncServerCA]
+	annos := map[string]string{
+		constants.ControllerConfigResourceVersionKey: controllerConfig.ObjectMeta.ResourceVersion,
+		constants.ControllerConfigSyncServerCA:       controllerConfig.Annotations[ctrlcommon.ServiceCARotateAnnotation],
+	}
+	if _, err := dn.nodeWriter.SetAnnotations(annos); err != nil {
```

Comment:
So this is setting this to true for the one sync then the next sync it sets it back to false if the controllerconfig one was reset?

(should we add a conditional that `ControllerConfigSyncServerCA` is not equal between node and controllerconfig before doing a write?)

Reply:
yeah. this keeps track of the last observed value of the cconfig one to avoid unnecessary kubelet restarts. I will add that check here.